opensafely/jupyter.py

```diff
@@ -163,19 +169,28 @@ def main(directory, name, port, no_browser, unknown_args):
         f"--name={name}",
         f"--hostname={name}",
         "--label=opensafely",
-        # note: on windows this will preserve drive letter, but switch to unix
-        # separators, which is what docker understands
-        f"-v={directory.resolve().as_posix()}:/workspace",
+        # note: // is to avoid git-bash path translation
+        f"-v={directory.resolve()}://workspace",
         "ghcr.io/opensafely-core/python",
     ]
 
-    # on windows, we need to wrap command in winpty to for docker run -it
-    winpty = shutil.which("winpty")
-    if winpty is not None:
-        debug(f"found winpty at {winpty}")
+    # git-bash's default terminal does not support a unix tty directly. So, try
+    # and detect this situation and wrap it ourselves
+    tty = (
```

Comment:
nit: I'd probably order these assignments as in the `if` below. The logical order does feel like should be in that `if`, that is:

* is this Windows?
* is winpty available?
* is this running in a tty?

You could also maybe skip the other checks entirely if Windows isn't running. But these look very inexpensive, so not a big problem either way.

Reply:
I've refactored into a function to make this cleaner.